splice/queries.py

```diff
@@ -358,6 +367,123 @@ def get_all_distributions(limit=100):
     return channels
 
 
+def get_upcoming_distributions(limit=100, leniency_minutes=15, include_past=False):
+    """
+    Obtain distributions, partitioned by channels with up to ``limit`` results
+    for each channel
+    :leniency_minutes: have a leniency in minutes up to the present when looking for distributions
+    """
+    from splice.environment import Environment
+
+    env = Environment.instance()
+
+    # getting around PEP8 E712 warning. This is necessary for SQLAlchemy
+    false_value = False
+
+    dist_cte = (
+        env.db.session
+        .query(
+            Distribution.id,
+            Distribution.channel_id,
+            Distribution.url,
+            Distribution.created_at,
+            Distribution.scheduled_start_date,
+            func.row_number().over(
+                partition_by=Distribution.channel_id,
+                order_by=Distribution.scheduled_start_date.asc())
+            .label('row_num')
+        )
+        .filter(Distribution.deployed == false_value))
+
+    if not include_past:
+        min_dt = datetime.utcnow() - timedelta(minutes=leniency_minutes)
+        dist_cte = (
+            dist_cte
+            .filter(Distribution.scheduled_start_date >= min_dt))
+
+    dist_cte = dist_cte.cte()
+
+    stmt = (
+        env.db.session
+        .query(
+            dist_cte.c.id,
+            dist_cte.c.channel_id,
+            dist_cte.c.url,
+            dist_cte.c.created_at,
+            dist_cte.c.scheduled_start_date)
+        .filter(dist_cte.c.row_num <= limit)
+        .order_by(dist_cte.c.scheduled_start_date.asc())
+    )
+
+    rows = stmt.all()
+
+    channels = {}
+
+    for row in rows:
+        c_dists = channels.setdefault(row.channel_id, [])
+        c_dists.append({'id': row.id, 'url': row.url, 'created_at': row.created_at, 'scheduled_at': row.scheduled_start_date})
+
+    return channels
+
+
+def get_scheduled_distributions(minutes, dt_query=None):
+    """
+    Returns distributions scheduled from a point in time, and a leniency period
+    within which a tasks could've been scheduled closed to that point.
+    As a regular task, it is intended to run at least once hourly.
+    :minutes: amount of time in the past from the query time which is still viable
+    :dt_query: optionally set the date time to find schedules for
+    """
+    from splice.environment import Environment
+
+    env = Environment.instance()
+
+    if not minutes or not (0 < minutes < 60):
+        raise ValueError("minutes needs to be a number between 1..59 inclusive")
```

Comment:
Stated as it is in the code, 1..59 is correct.

Arguably, a 0 time difference could be ok too